### PR TITLE
Redact passwords

### PR DESCRIPTION
**Pull request by DavidjohnBlodgett**
resolves RackHD/RackHD#241

@RackHD/corecommitters

new PR at the request of @brianparry 's to place logic into graph-object model.

due to a number of issues hit attempting to make my earlier code testable, I refactored to use encryption so passwords are not saved in our database in plain text to begin with and will be naturally concealed from the user on a query.

unit tests are complete for this feature.

### CODE REVIEW

File: lib/models/graph-object.js [hunk 1]
```diff
@@ -43,6 +43,38 @@ function GraphModelFactory (Model, Constants, configuration) {
...
+                                        break;
+                                    }
+                                }
+                                if ( !alreadyFound ) {
+                                    walked.push( workflow[ property ]);
+                                    stack.push( { workflow : workflow[ property ], stack: item.stack + '.' + property });
+                                }
+                            } else {
+                                if ( property.match( /password/i )) {
+                                    workflow[property] = '*** REDACTED ***';
```
Comment (yyscamper) [inline lib/models/graph-object.js]: I hope we can use the same redacted word, see: https://github.com/RackHD/on-core/blob/master/lib/serializables/log-event.js#L193

File: lib/models/graph-object.js [hunk 2]
```diff
@@ -43,6 +43,38 @@ function GraphModelFactory (Model, Constants, configuration) {
             active: function() {
                 var obj = this.toObject();
                 return Constants.Task.ActiveStates.indexOf(obj._status) > -1;
+            },
+            toJSON: function () {
```
Comment (yyscamper) [inline lib/models/graph-object.js]: I don't see the `toJSON` is called in API, so you need submit another PR to make sure the `toJSON` is called.
- https://github.com/RackHD/on-http/blob/a2da546c01dd7176a01f6bc93e8162e5bd1797e8/lib/api/1.1/northbound/workflows.js#L147
- https://github.com/RackHD/on-http/blob/a2da546c01dd7176a01f6bc93e8162e5bd1797e8/lib/api/2.0/workflows.js#L59

Comment (DavidjohnBlodgett) [inline lib/models/graph-object.js]: @yyscamper so Felix, the toJSON is an override of the base function in waterline for toJSON, so this is actually invoked automatically when a query is made (i've tested this)... however... I cannot have my cake and eat it too... I am unable to make unit tests using an override function like this, because of scope issues... I'll have to revert to making a class function and calling it in on-http like you suggest.

File: lib/models/graph-object.js (hunk 1 above)
Comment (DavidjohnBlodgett) [inline lib/models/graph-object.js]: @yyscamper ok, so you would prefer "[REDACTED]" vs "*** REDACTED ***" ? just making sure I understand the request.

File: spec/lib/models/graph-object-spec.js [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 var base = require('./base-spec');
 var  sandbox = sinon.sandbox.create();
+var graphObj;
```
Comment (DavidjohnBlodgett) [inline spec/lib/models/graph-object-spec.js]: oops missed removing this...

**Comment by DavidjohnBlodgett** [comment]
so the most recent changes, we have gone with encrypting the password property, this way we can keep it encrypted on disk when a graph is shipped to a collection and not have an expensive recursive search done with each api call/query. new output looks like this: ```                "repo": "{{api.server}}/esxi/{{options.version}}",
                "rootPassword": "ezgZbdAHMhU9+KViV5A4nA==.hSP54vlC8tvWWVhS37yb9m8elNOW2TLIeMyDpSGrZJU=",
                "users": [
                    {
                        "name": "onrack",
                        "password": "FDWcWfw054YE9qWFmE87Ug==.6X+GyHREqap6VsSE9iuAi+BDXb6BWCc/VhImrgztwWQ=",
                        "uid": 1010
                    }
                ],```

**Comment by DavidjohnBlodgett** [comment]
@brianparry so, I was wondering if I could get a second set of eye's on this (with ben being on vaca) so I'm worried, since this now encrypts passwords in the database... could other areas of our code expect plain text? and if that was the case... do these errors in jenkins seem to align with that thought process? thanks for any insight you can give!

**Comment by amymullins** [comment]
test this please

File: lib/models/graph-object.js (hunk 1 above)
Comment (yyscamper) [inline lib/models/graph-object.js]: no preference for me, just want to keep them using same value

File: lib/common/sanitizer.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+     */
+    return function Sanitizer (workflow) {
+        //workflow = JSON.parse(workflow);
+        var sanitizedworkflow = workflow;
+        var walked = [];
+        var stack = [ { workflow : workflow , stack : '' } ];
+        while ( stack.length > 0 ) {
+            var item = stack.pop();
+            var workflow  = item.workflow;
+            for ( var property in workflow ) {
```
Comment (brianparry) [inline lib/common/sanitizer.js]: Minor comment: lodash `forOwn` could simplify this logic a bit.

File: lib/common/sanitizer.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+                                alreadyFound = true;
+                                break;
+                            }
+                        }
+                        if ( !alreadyFound ) {
+                            walked.push( workflow[ property ]);
+                            stack.push( { workflow : workflow[ property ], stack: item.stack + '.' + property });
+                        }
+                    } else {
+                        if ( property.match( /password/i ) || property.match(/community/i)) {
```
Comment (brianparry) [inline lib/common/sanitizer.js]: There is a constant array of redaction patterns you can use for this. See https://github.com/RackHD/on-core/blob/master/lib/models/obms.js#L36

File: lib/common/sanitizer.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     *
+     */
+    return function Sanitizer (workflow) {
+        //workflow = JSON.parse(workflow);
+        var sanitizedworkflow = workflow;
+        var walked = [];
+        var stack = [ { workflow : workflow , stack : '' } ];
+        while ( stack.length > 0 ) {
```
Comment (brianparry) [inline lib/common/sanitizer.js]: General question on this function: rather than pushing object properties onto a stack, would it be easier to use a recursive function?  Maybe something like: 
```
function encryptObj(obj) {
    _.forOwn(obj, function(value, key) {
        if (typeof(value) === 'object') {
            encryptObj(value);
        } else if (key.match(secretsPattern) {
            obj[key] = encrypt(value);
        }
    }
}
```

File: lib/workflow/stores/mongo.js [hunk 7]
```diff
@@ -202,7 +203,7 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             new: true,
             upsert: true
         };
-
+        definition = sanitizer(definition);
```
Comment (brianparry) [inline lib/workflow/stores/mongo.js]: @DavidjohnBlodgett Could you move the sanitizer call into `findAndModifyMongo`?

**Comment by brianparry** [comment]
Thanks @DavidjohnBlodgett!  :+1: aside from the unit test failure.

**Comment by yyscamper** [comment]
Two questions:
- Since you encrypt the options while persisting graph object, so how and when to decrypt the password before if the job wants to get the real value?
- Output the ciphertext in API response still have security concern, since RackHD is open source, so it's clear that how to decrypt it. So this link redaction still leak the sensitive information.

**Comment by brianparry** [comment]
@yyscamper For obms, I added some convenience methods in the model to decrypt, but ultimately this should be the responsibility of the job that uses the secret.
On your second point, the problem is not that we're revealing the cipher; it's that we're storing our keys insecurely.  Really, we should be using something like this: https://www.hashicorp.com/blog/vault.html

**Comment by DavidjohnBlodgett** [comment]
@yyscamper hey Felix, so I believe your first question is the root problem for why some of the integration tests are currently failing... we are not decrypting the password before it is being used.... so yes, this is an open question, I am assuming I'll have to hunt down the locations we need to decrypt and add those to the PR. For the short term I would agree with @brianparry to have the features that use passwords be expected to decrypt.

**Comment by DavidjohnBlodgett** [comment]
@yyscamper @brianparry I am open to any suggestions (and ripping this logic out if we add database tagging), and can look into the vault option as a perhaps long term solution. (at this time, I was focused on a short term fix for some blocking ODR stuff and this seemed like a more elegant method of masking our sensitive data then just transforming an object from a database that's storing plaintext ) I do however understand that it isn't entirely secure... just better then what we got lol

**Comment by DavidjohnBlodgett** [comment]
@brianparry just a heads up, I implemented most of what you asked for... still looking at shifting up the sanitizer function call into findAndModifyMongo~ (having a nubby moment). In addition, maybe seeking some help sorting out what jobs exactly are failing that could lead to where we expect the passwords to be decrypted...

**Comment by pscharla** [comment]
test this please

**Comment by DavidjohnBlodgett** [comment]
I think I've tracked down this jenkins failure to 
```
Task.Base.Linux.Commands
```

working on a fix

**Comment by tldavies** [comment]
test this please

File: lib/workflow/stores/mongo.js [hunk 8]
```diff
@@ -202,7 +203,7 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
             new: true,
             upsert: true
         };
-
+        //sanitizer(definition);
```
Comment (yyscamper) [inline lib/workflow/stores/mongo.js]: Remove this line if you don't want it.

**Comment by yyscamper** [comment]
@DavidjohnBlodgett : 
The biggest problem of the PR is that it lacks the peer function of `santize`, the data is encrypted but it is not decrypted in job, so that if your PR is merged, I can think of there is a lot jobs will not work well, such as install-os.

**Comment by DavidjohnBlodgett** [comment]
@yyscamper I cannot agree more, I've been trying to wrap my brain around all possible effected areas. If you can think of any other area's we may need to unencrypt aside from OS installs and maybe ssh command tasks... That would be very helpful... I'm assuming I'll have to post a dependent PR in on-tasks to implement to decrypting stage for each specific job... unless there are any possible layers of abstraction I am unaware of that decryption could reside. There's a part of me that feels we may have bitten off more then we can chew, and may need to revert back to simple redaction at the api layer if this is a blocking issue for any internal group...

**Comment by DavidjohnBlodgett** [comment]
please don't merge yet... manual testing has found this new logic seems to miss a rootPassword property in my example, thats found outside of a workflows options... investigating now...

**Comment by DavidjohnBlodgett** [comment]
@RackHD/corecommitters ok guys... I think this is it... we made some changes to insure that decryption happens when we actually use it by tasks... and I manually tested that the recursion and regex work. This should be merge ready. Let me know if you have any questions, or concerns!

**Comment by rolandpoulter** [comment]
👍 although you could squash the commit history down a little.

**Comment by DavidjohnBlodgett** [comment]
@rolandpoulter understood.

**Comment by pscharla** [comment]
@RackHD/corecommitters Need merge for this one to resolve the P1 ODR for upcoming release.

**Comment by brianparry** [comment]
Looks good to me :+1:
